### DISCUSSION TITLE
[QUESTION] How to federation ?

### DISCUSSION THREAD

**Discussion by esakarya** [rancher/rke2#3215] [General]
Anyone can explain me , i need build cluster both datacenters. 
Please tell me how can i build kubernetes federation both datacenters.

**Comment by brandond** [comment]
Do you have any questions about how to do this with RKE2 specifically, or are you asking how Kubernetes federation works? Anything that works with any other Kubernetes distro should work with RKE2 - for example, KubeFed: https://github.com/kubernetes-sigs/kubefed

This is not something we would document as part of RKE2, however.

**Comment by esakarya** [comment]
Is there another way to add cluster nodes from a different datacenter?

**Reply by brandond** [reply]
I don't know anything about your data centers or their constraints. Have you tried? What happened?

As long as all the nodes can reach each other on the correct ports, and have access to the same underlying infrastructure (storage, for example), Kubernetes doesn't care where your nodes are physically located.

**Comment by esakarya** [comment]
Here is the error. one virtual machine is on azure another one is on cloud google. all ports accepted with firewall.

`Aug 05 12:58:53 node-1 rke2[3380]: time="2022-08-05T12:58:53Z" level=error msg="Failed to check local etcd status for learner management: context deadline exceeded"
Aug 05 12:58:53 node-1 rke2[3380]: time="2022-08-05T12:58:53Z" level=info msg="Waiting to retrieve kube-proxy configuration; server is not ready: https://127.0.0.1:9345/v1-rke2/readyz: 500 Internal Server Error"
Aug 05 12:58:54 node-1 rke2[3380]: time="2022-08-05T12:58:54Z" level=info msg="Waiting for etcd server to become available"
Aug 05 12:58:58 node-1 rke2[3380]: time="2022-08-05T12:58:58Z" level=info msg="Waiting to retrieve kube-proxy configuration; server is not ready: https://127.0.0.1:9345/v1-rke2/readyz: 500 Internal Server Error"
Aug 05 12:59:03 node-1 rke2[3380]: time="2022-08-05T12:59:03Z" level=info msg="Waiting to retrieve kube-proxy configuration; server is not ready: https://127.0.0.1:9345/v1-rke2/readyz: 500 Internal Server Error"
Aug 05 12:59:04 node-1 rke2[3380]: {"level":"warn","ts":"2022-08-05T12:59:04.223Z","logger":"etcd-client","caller":"v3@v3.5.4-k3s1/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc000c5c540/127.0.0.1:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = latest balancer error: last connection error: connection error: desc = \"transport: Error while dialing dial tcp 127.0.0.1:2379: connect: connection refused\""}
`

**Comment by brandond** [comment]
> one virtual machine is on azure another one is on cloud google.

All of the servers running etcd need to be able to reach each other at their **private** IP addresses. I suspect you do not have a VPN or other tunnel in place that allows the azure and google nodes to connect directly to each other. In this scenario you would need to have all of your etcd nodes on the same provider, and only spread the agents across the two.

I would also note that most of the CNIs use VXLAN to tunnel cluster traffic between nodes. VXLAN is not encrypted, so any traffic transiting the internet would not be secure.
